Add LinkHelper class

crobert-betterdoc opened the pull request:
The LinkHelper is designed to be used by any containerservice template (or other code) that needs to link to a separate stack so we provide this as a global helper available to all templates automatically.

Parameter values are automatically escaped so that the generated URL created by the LinkHelper are valid and do not create any problems when opened within a browser.

Commit:
Add LinkHelper class

The LinkHelper is designed to be used by any containerservice template (or other code) that needs to link to a separate stack so we provide this as a global helper available to all templates automatically.

Parameter values are automatically escaped so that the generated URL created by the LinkHelper are valid and do not create any problems when opened within a browser.

Commit:
Fix Rubocop violations

vlado commented on lib/betterdoc/containerservice/helpers/link_helper.rb:
```diff
@@ -0,0 +1,45 @@
+          result_url
+
+        end
+        # rubocop:enable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
```
Just a quick idea how to refactor this:

```ruby
def create_stacker_link(target_path, parameters = {})
  base_url = request.headers['HTTP_X-STACKER-ROOT-URL'].presence || ENV['STACKER_ROOT_URL'].presence || "/"
  [URI.join(base_url, target_path), parameters.to_query].select(&:present?).join("?")  
end
```

This is just a quick idea, you can further improve and/or play with https://docs.ruby-lang.org/en/2.1.0/URI.html if you like. `to_query` is Rails method https://api.rubyonrails.org/classes/Hash.html#method-i-to_query

Also, since it returns url I would maybe rename this method to something more rails like `stacker_url_for(...)` and then add other method called `stacker_link_to(text, url, options)` that will delegate to standard `link_to` helper but will pregenarate url.

```ruby
def stacker_link_to(text, url, html_options = {})
  link_to(text, stacker_url_for(url), html_options)
end
``` 

Only problem is how to pass parameters in that case. Could be maybe done like this.
```ruby
def stacker_link_to(text, url, html_options = {})
  parameters = html_options.delete(:parameters)
  link_to(text, stacker_url_for(url, parameters), html_options)
end
```

crobert-betterdoc replied:
Good idea, will take a look at this!

vlado commented on README.md:
````diff
@@ -80,6 +80,50 @@ Renders the template specified under `:template_name`.
+
+```html
+<div>
+    <%= create_stacker_link('another/stack/location') %>
````
This will return only url, and I think `<a>` tag is actually expected.

crobert-betterdoc replied:
A, good catch!

Commit:
Make documentation match actual link creation

The create_stacker_link method only creates the URL and not the complete a tag.

Commit:
Simplify base_url computation

Commit:
Simplify link generation and adjust naming

Use "stacker_link_url" instead of "create_stacker_link" to signalize that not the link itself is being created by the helper but "only" the URL that can be used as "href" attribute of the link. Also adjust the documentation accordingly.

Commit:
Rails always uses underscores to replace "-" characters

Commit:
Add missing ".presence" to check of environment variable

vlado commented:
Everything looks fine now :+1:
